proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,319 @@
+   quick feedback to the user. Any module imports of a module using
+   ``TemplateHaskell`` must be compiled to object code.
+   This is despite the fact that we will not generate object code for the module
+   itself. By distinguishing imported modules whose code is executed only at
+   runtime, we can reduce this unfortunate work significantly, and entirely in many
+   cases.
+3. Projects such as haskell-language-server face similar problems as 2., where they are interested only in the result of type-checking modules, but when ``TemplateHaskell`` is enabled a large
+   number of modules have to be cautiously compiled to bytecode.
+4. By using splice imports we can separate the dependencies into those only needed at build-time (1) and
+   those only needed at runtime (2). We can then link only against those packages needed at runtime.
+5. Currently, when cross-compiling, in modules that use ``TemplateHaskell``, all
+   imported modules must be compiled for both host and target.
+   By distinguishing imported modules used at compile time(i.e. not used at
+   runtime), we can require only those modules to be compiled for the host.
+   Similarly, by distinguishing imported modules used at runtime (i.e. not used at
+   compile time), we can require only those modules to be compiled for the
+   target. It can be very hard or impossible to make some packages available on
+   some cross-compile target platforms, so this change would significantly
+   improve the applicability of ``TemplateHaskell`` in these scenarios.
+
+
+Definitions
+-----------
+
+level
+  Each expression exists at a level. The level is increased by 1 when
+  inside a quote and decreased by 1 inside a splice. Therefore the level of
+  an expression can be calculated as the number of quotes surrounding an expression
+  subtract the number of splices. For example::
+
+    -- foo is at level 0
+    foo = $(let
+      -- bar is at level -1
+      bar = $(let
+        -- baz is at level -2
+        baz = [|
+        -- qux is at level -1
+          qux = [|
+            -- quux is at level 0
+            quux = [|
+              quuz is at level 1
+              quuz = 0
+            |]
+          |]
+        |] in baz
+      ) in bar
+    )
+
+
+top-level splice
+  A splice, where the body is at a negative level, an unadorned
+  declaration splice or a quasiquoter.
+
+
+home module
+  A module from the package that is currently being compiled.
+
+
+Proposed Change
+---------------
+
+The ``splice`` modifier indicates that a module's imports are available at compile time.
+The absence of the ``splice`` modifier indicates that a module's imports are available
+at runtime.
+
+A ``splice`` imported module allows identifiers to be used at compile time. Compile time
+evaluation is indicated by a top-level splice. Therefore the ``splice`` modifier
+affects name resolution depending on whether we are inside a top-level splice or not.
```

Comment:
This bit about name resolution is quite worrisome. Let’s say I write

```
import Control.Lens

makeLenses ‘’Foo
```

Do I get an error that `makeLenses` is not in scope?

```
Test.hs:5:1: error:
    Variable not in scope:
      makeLenses
        :: Language.Haskell.TH.Syntax.Name
           -> Language.Haskell.TH.Lib.Internal.DecsQ
```

That’s not a good UX. I think there’s an alternative. Do name resolution as usual, brining all names into regardless of the splice level. But when executing the splice, issue an error if the name isn’t splice-imported:

```
Test.hs:5:1: error:
    ‘makeLenses’ is not available in top-level splices.
    Did you mean to use a splice-import?
```

Reply:
Yes, that is the proposed behavior. Updated now to make it clearer.